crates/miden-lib/asm/miden/contracts/faucets/mod.masm

```diff
@@ -0,0 +1,120 @@
+#! - tag is the tag to be included in the note.
+#! - aux is the auxiliary data to be included in the note.
+#! - note_type is the type of the note that holds the asset.
+#! - execution_hint is the execution hint of the note that holds the asset.
+#! - RECIPIENT is the recipient of the asset, i.e.,
+#!   hash(hash(hash(serial_num, [0; 4]), script_root), input_commitment).
+#!
+#! Panics if:
+#! - the transaction is being executed against an account that is not a fungible asset faucet.
+#! - the total issuance after minting is greater than the maximum allowed supply.
+#!
+#! Invocation: exec
+export.distribute
+    # get max supply of this faucet. We assume it is stored at pos 3 of slot 1
+    push.METADATA_SLOT exec.account::get_item drop drop drop
+    # => [max_supply, amount, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # get total issuance of this faucet so far and add amount to be minted
+    exec.faucet::get_total_issuance
+    # => [total_issuance, max_supply, amount, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # compute maximum amount that can be minted, max_mint_amount = max_supply - total_issuance
+    sub
+    # => [max_supply - total_issuance, amount, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # check that amount =< max_supply - total_issuance, fails if otherwise
+    dup.1 gte assert.err=ERR_FUNGIBLE_ASSET_DISTRIBUTE_WOULD_CAUSE_MAX_SUPPLY_TO_BE_EXCEEDED
+    # => [amount, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # creating the asset
+    exec.faucet::create_fungible_asset
+    # => [ASSET, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # mint the asset; this is needed to satisfy asset preservation logic.
+    exec.faucet::mint
+    # => [ASSET, tag, aux, note_type, execution_hint, RECIPIENT]
+
+    # store and drop the ASSET
+    movdnw.2
+    # => [tag, aux, note_type, execution_hint, RECIPIENT, ASSET]
+
+    # create a note
+    exec.output_note::create
+    # => [note_idx, ASSET]
+
+    # load the ASSET and add it to the note
+    movdn.4 exec.output_note::add_asset
+    # => [ASSET, note_idx]
+
+    dropw drop
+    # => []
+end
+
+#! Burns the fungible asset from the active note.
+#!
+#! This procedure retrieves the asset from the active note and burns it. The note must contain
+#! exactly one asset, which must be a fungible asset issued by this faucet.
+#!
+#! Inputs:  [pad(16)]
+#! Outputs: [pad(16)]
+#!
+#! Where:
+#! - ASSET is the fungible asset that was burned.
+#!
+#! Panics if:
+#! - the procedure is not called from a note context (active_note::get_assets will fail).
+#! - the note does not contain exactly one asset.
+#! - the transaction is executed against an account which is not a fungible asset faucet.
+#! - the transaction is executed against a faucet which is not the origin of the specified asset.
+#! - the amount about to be burned is greater than the outstanding supply of the asset.
+#!
+#! Invocation: call
+export.burn
```

Comment:
Nit: Since this procedure is `call`-ed, it would be good to add `pad`s in the stack comments.